### DISCUSSION TITLE
Way to measure processing time?

### DISCUSSION THREAD

**Discussion by CROprogrammer** [sngular/kloadgen#398] [Q&A]
Hi,

is there any way to measure how long it has taken for a massage to be processed in Kafka processor using KLoadGen?

I'm thinking something like how long it took for produced message that was written to input topic, and then processed in Kafka processor, and finally written to output topic. How could I measure how long it took for specific message when it was produced, to when it was consumed?
I'm aware that KLoadGen can produce and consume messages, but can it measure time between those two action.

**Comment by jemacineiras** [comment]
Hi @CROprogrammer,

That can be a bit tricky but doable. There is a consumer sampler in Kloadgen, so the best approach here is hace producer and consumer in the test plan each one to the required topics. We populate the produced and received message into the JMeter context so here your goal is build and assertion which got the information when those messages where populated (produced and received) so you can the timestamp and obtain the difference.

We do not have an example about, so this is more or less a theoretically answer.

**Reply by CROprogrammer** [reply]
Thank you, I will try to make it work when https://github.com/sngular/kloadgen/issues/395 gets solved

**Reply by CROprogrammer** [reply]
How do you access JMeter context so that I could read produced and received messages from it?

**Reply by CROprogrammer** [reply]
And how can I get time when messages was produced, and consumed from context?
Could you provide some example by any chance

**Reply by jemacineiras** [reply]
Hi there, 

Nowadays we don´t have a working example, since this is pure JMeter functionality, but we will consider generate one and added to your example test plan.

For now the best approach is use the JSR223 components and access the SampleResult object like in [this](https://jmetervn.com/2016/12/20/jsr223-with-groovy-variables-part-2/) example.

In the produce case, the SampleResult object we add the produced message into the Sample Data field and response from Kafka Cluster in the Response Data.
In the consumer case, we use the ReponseData field to return the consumed message.

Wish that help you.

**Reply by CROprogrammer** [reply]
I've made Producer, Consumer, and I have Kafka Streams processor that receives messages from KLoadGen Producer, waits for 5 seconds and then writes slightly (content-wise) changed message to output topic.
When Producer produces first message, my processor works fine, and writes new message to the output topic. But the thing is that KLoadGen Consumer only consumes that message after I've run it again. I expect Consumer to wait for message to be written to the output topic from which it consumes messages rather then that I have to run KLoadGen again for it to work. Is that even possible?
And because of that I'm unable to pinpoint when message was actually consumed, and because of that processing time is unknown.

**Reply by CROprogrammer** [reply]
Is it possible for Consumer to work all the time paralley with the Producer, because my Producer works inifitely, but Consumer stops after consuming one message?

**Reply by jemacineiras** [reply]
Hi,
Consumer Sampler have a loop which waits until message is read or timeout, so it you added in the same Thread Group than the publisher you will block it. Ideally it should be run in its own Thread Group and in parallel as producer. 
By default timeout is 5000 millis, I think is time moire than enough to get the message you produce in your server. I that approach does not work I will need to see your test plan file.

Cheers

**Reply by CROprogrammer** [reply]
Yeah those 5000 ms were my problem, my processor waited for 5 seconds and because of that it did not work quite right

**Reply by jemacineiras** [reply]
Timeout can be set up in the Sampler properties, under `timeout.millis`
